src/Monkeypatcher.cc

```diff
@@ -737,6 +954,115 @@ bool Monkeypatcher::try_patch_syscall(RecordTask* t, bool entering_syscall) {
   return true;
 }
 
+bool Monkeypatcher::try_patch_syscall_aarch64(RecordTask* t, bool entering_syscall)
+{
+  Registers r = t->regs();
+  remote_code_ptr ip = r.ip() - 4;
+
+  uint32_t inst[2] = {0, 0};
+  size_t bytes_count = t->read_bytes_fallible(ip.to_data_ptr<uint8_t>() - 4, 8, &inst);
+  if (bytes_count < sizeof(inst) || inst[1] != 0xd4000001) {
+    LOG(debug) << "Declining to patch syscall at "
+               << ip << " for unexpected instruction";
+    tried_to_patch_syscall_addresses.insert(ip);
+    return false;
+  }
+  // mov x8, 0xdc
+  if (inst[0] == 0xd2801b88) {
+    // Clone may either cause the new and the old process to share stack (vfork)
+    // or replacing the stack (pthread_create)
+    // and requires special handling on the caller.
+    // Our syscall hook cannot do that so this would have to be a raw syscall.
+    // We can handle this at runtime but if we know the call is definately
```

Comment:
"definitely"

What's the point of this code? We do need to handle this at runtime somehow so we why is it worth also handling it here?

Reply:
I feel like it's still cleaner to handle it here and I assumed that if quite free to check since we are already reading tracee memory anyway. I originally thought that I have to make the check in the stub and make the syscall there which would mean that the clone will be issues without a usable backtrace but since I am actually making it in librrpreload.so that's actually not a problem anymore.

This does match what x86 does better though. The fork/vfork/clone actually never get patched on x86 AFAICT. This is the main reason I decided to keep it even after I have a runtime fix for the general case...

Reply:
Why aren't fork/vfork/clone patched on x86? We just don't have a syscall hook for the code pattern they use?

IMHO having two ways to handle clone is not cleaner than having only one way to handle it.

Reply:
AFAICT the code pattern for those aren't handled (actually I'm less sure about fork but I'm pretty sure about clone and vfork).

I was trying to see what x86 does and I realized that the syscall site didn't get patched.... On my laptop (archlinux x86_64) `clone` has `syscall` followed by `test %rax,%rax` and conditional branch whereas `vfork` has the syscall followed by `push %rdi`. None of these seems to be covered by the patterns.

I guess clean isn't the right word. I'm mainly saying that for the most common case (I assume few code use syscall to call clone) if we know the call is almost certainly something that will skip syscallbuf, we can just avoid the patch to keep perturbation on the callee to be minimum. Given that this path isn't tested on x86, I was also worrying a little if there's any other side effects that I haven't considered.....